verifier/src/main/java/com/exasol/classlistextractor/verifier/ClassListServer.java

```diff
@@ -0,0 +1,184 @@
+package com.exasol.classlistextractor.verifier;
+
+import java.io.*;
+import java.net.*;
+import java.nio.charset.StandardCharsets;
+import java.util.*;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import com.exasol.errorreporting.ExaError;
+
+import lombok.Getter;
+
+/**
+ * This class receives the class list from the agent running in the UDFs via TCP.
+ */
+class ClassListServer implements AutoCloseable {
+    private final ClientHandler clientHandler;
+
+    /**
+     * Create a new instance of {@link ClassListServer}.
+     */
+    public ClassListServer() {
+        this.clientHandler = new ClientHandler();
+    }
+
+    /**
+     * Get the port the server listens on.
+     * 
+     * @return port number
+     */
+    public int getPort() {
+        return this.clientHandler.getPort();
+    }
+
+    @Override
+    public void close() {
+        this.clientHandler.stop();
+    }
+
+    /**
+     * Clear the captured class list.
+     */
+    public void clear() {
+        this.clientHandler.clear();
+    }
+
+    /**
+     * Get the received class list.
+     * <p>
+     * The class list is a union (with duplicate removal) of all revieved class lists since the last call to
+     * {@link #clear()}.
+     * </p>
+     * 
+     * @return list of classes.
+     */
+    public List<String> getClassList() {
+        return this.clientHandler.getClassList();
+    }
+
+    private static class ClientHandler implements Runnable {
+        private final ServerSocket serverSocket;
+        private final AtomicBoolean isRunning = new AtomicBoolean(true);
+        private final ConcurrentLinkedQueue<Future<List<String>>> classListReaders = new ConcurrentLinkedQueue<>();
+        @Getter
+        private final int port;
+        private final ServerThreadMonitor serverThreadMonitor = new ServerThreadMonitor();
+
+        ClientHandler() {
+            try {
+                this.serverSocket = new ServerSocket(0);
+                this.serverSocket.setSoTimeout(50);
+                this.port = this.serverSocket.getLocalPort();
+            } catch (final IOException exception) {
+                throw new UncheckedIOException(ExaError.messageBuilder("E-JCLE-VF-3")
+                        .message("Failed to start a server for retrieving the class lists of a project.").toString(),
+                        exception);
+            }
+            final Thread thread = new Thread(this);
+            thread.start();
+        }
+
+        void stop() {
+            this.isRunning.set(false);
+            try {
+                this.serverSocket.close();
+            } catch (final IOException exception) {
+                throw new UncheckedIOException(ExaError.messageBuilder("E-JCLE-VF-6")
+                        .message("Failed to close the server for retrieving the class lists of a project.").toString(),
+                        exception);
+            }
+        }
+
+        synchronized void clear() {
```

Comment:
This is the only `synchronized` method. Did you mean to use a `synchronized (this.serverThreadMonitor) {}` block like in the other methods?

Reply:
good catch. It's was outdated.